src/main/java/org/rumbledb/compiler/TranslationVisitor.java

```diff
@@ -831,69 +751,57 @@ public Void visitObjectConstructor(JsoniqParser.ObjectConstructorContext ctx) {
         } else {
             Expression childExpr;
-            this.visitExpr(ctx.expr());
-            childExpr = this.currentExpression;
-            this.currentExpression = new ObjectConstructorExpression(
+            childExpr = (Expression) this.visitExpr(ctx.expr());
+            return new ObjectConstructorExpression(
                     childExpr,
                     createMetadataFromContext(ctx)
             );
         }
-
-        return null;
     }
 
     @Override
-    public Void visitArrayConstructor(JsoniqParser.ArrayConstructorContext ctx) {
+    public Node visitArrayConstructor(JsoniqParser.ArrayConstructorContext ctx) {
         if (ctx.expr() == null)
-            this.currentExpression = new ArrayConstructorExpression(createMetadataFromContext(ctx));
+            return new ArrayConstructorExpression(createMetadataFromContext(ctx));
         else {
-            this.visitExpr(ctx.expr());
-            Expression content = this.currentExpression;
-            this.currentExpression = new ArrayConstructorExpression(content, createMetadataFromContext(ctx));
+            Expression content = (Expression) this.visitExpr(ctx.expr());
+            return new ArrayConstructorExpression(content, createMetadataFromContext(ctx));
         }
-        return null;
-
     }
 
     @Override
-    public Void visitParenthesizedExpr(JsoniqParser.ParenthesizedExprContext ctx) {
+    public Node visitParenthesizedExpr(JsoniqParser.ParenthesizedExprContext ctx) {
         if (ctx.expr() == null)
-            this.currentExpression = new CommaExpression(createMetadataFromContext(ctx));
+            return new CommaExpression(createMetadataFromContext(ctx));
         else {
-            this.visitExpr(ctx.expr());
-            // implicit: this.currentExpression = this.currentExpression;
+            return this.visitExpr(ctx.expr());
         }
-        return null;
     }
 
     @Override
-    public Void visitVarRef(JsoniqParser.VarRefContext ctx) {
+    public Node visitVarRef(JsoniqParser.VarRefContext ctx) {
         String name = ctx.name.getText();
         if (ctx.ns != null)
             name = name + ":" + ctx.ns.getText();
-        this.currentExpression = new VariableReferenceExpression(name, createMetadataFromContext(ctx));
-        return null;
+        return new VariableReferenceExpression(name, createMetadataFromContext(ctx));
     }
 
     @Override
-    public Void visitContextItemExpr(JsoniqParser.ContextItemExprContext ctx) {
-        this.currentExpression = new ContextItemExpression(createMetadataFromContext(ctx));
-        return null;
+    public Node visitContextItemExpr(JsoniqParser.ContextItemExprContext ctx) {
+        return new ContextItemExpression(createMetadataFromContext(ctx));
     }
 
     @Override
-    public Void visitSequenceType(JsoniqParser.SequenceTypeContext ctx) {
+    public Node visitSequenceType(JsoniqParser.SequenceTypeContext ctx) {
```

Comment:
I have noticed that SequenceType is currently not a Node. Is this due to the XQuery spec?

If this was a node as well we could get rid of the last field in the TranslationVisitor.

Reply:
It is not (it is only a property of some expressions). However, I nevertheless was able to get rid of the field, via a util method.

Reply:
Very nice